llama_index/ingestion/pipeline.py

```diff
@@ -310,6 +311,45 @@ def _handle_upserts(self, nodes: List[BaseNode]) -> List[BaseNode]:
 
         return list(deduped_nodes_to_run.values())
 
+    def _handle_upserts_and_delete(self, nodes: List[BaseNode]) -> List[BaseNode]:
+        """Handle docstore upserts by checking hashes and ids, followed by deletion of missing documents in a docstore namespace."""
+        assert self.docstore is not None
+
+        existing_doc_ids_before = set(self.docstore.get_all_document_hashes().values())
+        doc_ids_from_nodes = set()
+        deduped_nodes_to_run = {}
+        for node in nodes:
+            ref_doc_id = node.ref_doc_id if node.ref_doc_id else node.id_
+            doc_ids_from_nodes.add(ref_doc_id)
+            existing_hash = self.docstore.get_document_hash(ref_doc_id)
+            if not existing_hash:
+                # document doesn't exist, so add it
+                self.docstore.add_documents([node])
+                self.docstore.set_document_hash(ref_doc_id, node.hash)
+                deduped_nodes_to_run[ref_doc_id] = node
+            elif existing_hash and existing_hash != node.hash:
+                self.docstore.delete_ref_doc(ref_doc_id, raise_error=False)
+
+                if self.vector_store is not None:
+                    self.vector_store.delete(ref_doc_id)
+
+                self.docstore.add_documents([node])
+                self.docstore.set_document_hash(ref_doc_id, node.hash)
+
+                deduped_nodes_to_run[ref_doc_id] = node
+            else:
+                continue  # document exists and is unchanged, so skip it
+
+        # Identify missing docs and delete them from docstore and vector store
+        doc_ids_to_delete = existing_doc_ids_before - doc_ids_from_nodes
```

Comment:
I think the majority of the code above is copy-paste from upsert

I wonder if it's fine to use the existing upsert function and then have an if statement for this new piece below

Reply:
(Sorry, should have finished reviewing this last night haha)

Reply:
I agree! I thought about it but didn't want to change _handle_upserts() to avoid any impact. I need to loop through nodes to keep a track of doc_ids in nodes. For that, I will have to change _handle_upserts() and return an extra variable. I didn't check if _handle_upserts() has been used anywhere else in the codebase. If it isn't, I can quickly make that change. What do you think?

Reply:
Sorry, I think I misunderstood your comment. Yes, I can incorporate this change in the existing _handle_upserts() method. Do you want me to do that?

Reply:
I can't think of an elegant way to combine this in _handle_upserts(). I might have to use multiple ifs if I want to avoid using existing_doc_ids_before and doc_ids_from_nodes in case of just upserts. or do you think it's okay to use them and ignore them for just upserts?

Reply:
Yea that's what I meant 👍🏻 It is only used here so it should be fine to modify slightly

Reply:
Basically

```
def _handle_upserts(...):
   <same code as before, but also keeping track of doc_ids>
   if self.docstore_strategy == upsert_and_delete:
      <your new code>
      
  return ...
```

Reply:
Done and tested, please review and let me know in case of any further comments. Thanks!